### ISSUE TITLE
Inconsistent with paper in terms of last activation function

### ISSUE THREAD

**Issue by hiyijian** [MSusik/Inhibited-softmax#1]
Dear @MSusik ,
Your paper claimed that cauchy PDF function, "1/(1+x^2)",  is used. However,  code of this repo uses "1 / (1 + 10x^2)" for MNIST and custom nonlinear function for CIFAR-10.
Do you have any further explain on this?

**Comment by MSusik** [comment]
Hi!

Thank you for your contribution! Indeed, the activation functions in the repo were different than in the publication.

I must apologise, it seems I pushed our development version of code where we tested few hypothesis. For a quick fix, I changed the activation function and checked if I can reproduce the results. They seem in par with the paper, so I pushed the change to the `master` branch.

Due to this fact, I suspect that you might bump into other inconsistencies somewhere in the code. On the weekend I will review the whole codebase and check if there are any other lines that should be changed. I will come back to you with what I found.

As a sidenote: Feel free to use the custom nonlinear function. It will give very similar results to the cauchy PDF, as shown in the appendix.

On the other hand, `1 / (1 + 10x^2)` is a bit tougher to train. We experimented with changing the variance in the activation function, but the results were not interesting, so they did not make it to the paper.

**Comment by hiyijian** [comment]
Thanks for your reply.
Maybe another inconsistence: According to  ablation study in the paper, the best performance is obtained with lamda of L2 penalty = 1e-4, But ISCifar-10 expriment uses 1e-3 in this repo. 1e-4 or 2e-4 or 4e-4 is widely adopted by deep learning community, 1e-3 maybe too big?
![image](https://user-images.githubusercontent.com/4326958/47057037-99a74580-d1f1-11e8-8974-9bbda39e6377.png)

**Comment by MSusik** [comment]
You are perfectly right that lower values of l2 are more often used. Moreover, in the paper and in the source code we use even stronger l2, namely 1e-2:

```
def is_loss(model):
    return
         lambda pred,y: cross_entropy(pred,y)
         + 0.01 * (model.dense3.weight.data ** 2).sum()
         + 0.000001 * pred.sum()
     )
```

and in the paper:

`In the case of Inhibited Softmax we set l2 penalty to 0.01,`

Note that we have a reason for that. In our work l2 penalty is used mostly as a tool for limiting the output of the softmax, while in the other papers it is used as a regularization term for overfitting prevention. We set it to so high a value in order to be sure that the OOD results are influenced by it. Obviously, one might argue that such a strong l2 penalty would deteriorate performance of larger networks. In ablation study we simply show that it is not that important to set it to such high values: it is important to have any penalty. We decided not to choose the best results after seeing the ablation study, as in our opinion it would be cherry-picking of the results. The ablation study was one of the last experiments we performed.

**Comment by hiyijian** [comment]
Well, @MSusik , I got. Thanks for the detail again

**Comment by MSusik** [comment]
@hiyijian I checked the repo. Our method is implemented as in the paper, but I noticed that it was impossible to reproduce all the sentiment experiments. I added the file, and added a script that automatically downloads all the datasets and puts them in the required directories.

Now the repo should be much more convenient, Let me know if you spot any other issues!

**Comment by hiyijian** [comment]
@MSusik Thanks. I will try your proposed method later and feedback to you
